File: src/learning_action_cpp/src/sum_action_server.cpp
/**
 * 需求：编写动作通信服务端 提取客户端发来的目标数字 然后从1开始遍历 将数字相加作为最终响应结果 并处理连续进度反馈
 * 步骤：
 *      1、包含头文件
 *      2、初始化 ROS2 客户端
 *      3、自定义节点类
 *          3-1、创建动作服务端
 *          3-2、提取客户端发来的目标 并作出回应
 *          3-3、处理客户端的取消请求 并作出回应
 *          3-4、生成连续反馈 并响应最终结果
 *      4、调用 spin 函数 并传入节点对象指针
 *      5、释放资源
 */

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "base_interfaces/action/sum_to_n.hpp"

using base_interfaces::action::SumToN;
using std::placeholders::_1;
using std::placeholders::_2;

class SumToNActionServer : public rclcpp::Node {
public:
    SumToNActionServer() : Node("sum_to_n_action_server_cpp") {
        RCLCPP_INFO(this->get_logger(), "动作服务端创建成功!");
        /*
            rclcpp_action::Server<ActionT>::SharedPtr
            create_server<ActionT, NodeT>(
                NodeT node, const std::string &name,
                rclcpp_action::Server<ActionT>::GoalCallback handle_goal,
                rclcpp_action::Server<ActionT>::CancelCallback handle_cancel, rclcpp_action::Server<ActionT>::AcceptedCallback handle_accepted
            )
        */
        server_ = rclcpp_action::create_server<SumToN>(
            this,
            "sum_to_n",
            std::bind(&SumToNActionServer::goal_callback, this, _1, _2),
            std::bind(&SumToNActionServer::cancel_callback, this, _1),
            std::bind(&SumToNActionServer::accepted_callback, this, _1)
        );
    }

private:
    /*
        std::function<GoalResponse(const GoalUUID &, std::shared_ptr<const typename ActionT::Goal>)>;
    */
    rclcpp_action::GoalResponse goal_callback(const rclcpp_action::GoalUUID &goal_uuid,
                                              std::shared_ptr<const SumToN::Goal> goal) {
        (void)goal_uuid;
        if (goal->num <= 1) {
            RCLCPP_WARN(this->get_logger(), "提交的数字必须大于1!");
            return rclcpp_action::GoalResponse::REJECT;
        }

        RCLCPP_INFO(this->get_logger(), "提交的数字合法!");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    /*
        std::function<CancelResponse(std::shared_ptr<ServerGoalHandle<ActionT>>)>;
    */
    rclcpp_action::CancelResponse cancel_callback(std::shared_ptr<rclcpp_action::ServerGoalHandle<SumToN>> server_goal_handle) {
        (void)server_goal_handle;
        RCLCPP_INFO(this->get_logger(), "收到取消请求, 允许取消!");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    /*
        std::function<void (std::shared_ptr<ServerGoalHandle<ActionT>>)>;
    */
    void accepted_callback(std::shared_ptr<rclcpp_action::ServerGoalHandle<SumToN>> server_goal_handle) {
        std::thread(std::bind(&SumToNActionServer::execute, this, server_goal_handle)).detach();
    }

    void execute(std::shared_ptr<rclcpp_action::ServerGoalHandle<SumToN>> server_goal_handle) {
        auto feedback = std::make_shared<SumToN::Feedback>();
        auto result = std::make_shared<SumToN::Result>();

        int num = server_goal_handle->get_goal()->num;
        int sum = 0;
        double progress = 0.0;

        /*
            void publish_feedback(std::shared_ptr<base_interfaces::action::SumToN_Feedback> feedback_msg)
        */
        rclcpp::Rate rate(1.0);
        for (int i = 1; i <= num; i++) {
            if (server_goal_handle->is_canceling()) {
                result->sum = sum;
                server_goal_handle->canceled(result);
                RCLCPP_INFO(this->get_logger(), "任务取消成功!");

                return;
            }

            sum += i;
            progress = i / (double)num;
            feedback->progress = progress;

            RCLCPP_INFO(this->get_logger(), "当前进度: %.2f", feedback->progress);
            server_goal_handle->publish_feedback(feedback);

            rate.sleep();
        }

        /*
            void succeed(std::shared_ptr<base_interfaces::action::SumToN_Result> result_msg)
        */
        if (rclcpp::ok()) {
            result->sum = sum;
            RCLCPP_INFO(this->get_logger(), "任务处理完毕, 最终结果为: %ld", result->sum);
            server_goal_handle->succeed(result);
        }
    }

    rclcpp_action::Server<SumToN>::SharedPtr server_;

};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<SumToNActionServer>());
    rclcpp::shutdown();

    return 0;
}

File: src/learning_action_py/learning_action_py/sum_action_server.py
"""实现整数累加动作服务端，处理目标、取消、进度反馈和最终结果."""
import time

import rclpy
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node

from base_interfaces.action import SumToN


class SumToNActionServer(Node):
    def __init__(self):
        super().__init__("sum_to_n_action_server_py")
        self.get_logger().info("动作服务端创建成功!(Python)")

        self.callback_group_ = ReentrantCallbackGroup()
        self.server_ = ActionServer(
            self,
            SumToN,
            "sum_to_n",
            self.execute_callback,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            callback_group=self.callback_group_,
        )

    def goal_callback(self, goal_request):
        if goal_request.num <= 1:
            self.get_logger().error("请提交大于1的数字!")
            return GoalResponse.REJECT

        self.get_logger().info("提交的数字合法!")
        return GoalResponse.ACCEPT

    def cancel_callback(self, cancel_request):
        self.get_logger().info("收到取消请求, 允许取消!")
        return CancelResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle):
        goal_handle.execute()

    def execute_callback(self, goal_handle):
        num = goal_handle.request.num
        total = 0

        feedback = SumToN.Feedback()
        result = SumToN.Result()

        for i in range(1, num + 1):
            if goal_handle.is_cancel_requested:
                result.sum = total
                goal_handle.canceled()

                self.get_logger().info("任务取消成功!")
                return result

            total += i

            feedback.progress = i / num
            goal_handle.publish_feedback(feedback)

            self.get_logger().info(f"持续反馈中: {feedback.progress:.2f}")
            time.sleep(1.0)

        result.sum = total
        goal_handle.succeed()

        self.get_logger().info(f"任务处理完成, 最终结果为: {result.sum}")
        return result


def main():
    rclpy.init()

    server = SumToNActionServer()
    executor = MultiThreadedExecutor()
    executor.add_node(server)

    try:
        executor.spin()
    except KeyboardInterrupt:
        pass
    finally:
        executor.shutdown()
        server.destroy_node()

        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
